designs/generics-design.md

```diff
@@ -0,0 +1,403 @@
+```
+int --> 0
+bool --> False
+class --> None
+```
+
+---
+
+## Proposed changes to the Parser
+
+### Type Variable Definitions
+
+Parser should identify and parse type-variable definitions as part of the global definitions.
+
+```typescript
+export function traverseDefs(c : TreeCursor, s : string) : [Array<VarInit<null>>, Array<FunDef<null>>, Array<Class<null>>, Array<TypeVar<null>>] {
+  ... 
+  const typevars: Array<TypeVar<null>> = [];
+
+  while(true) {
+    ...
+    else if(isTypeVar(c, s) {
+      typevars.push(traverseTypeVar(c, s));  
+    } else {
+      return [inits, funs, classes, typevars];
+    }
+    c.nextSibling();
+  }
+}
+```
+
+### Type Annotations of Generic Types
+
+Parser should be able to pass fancy type annotations of generic types with type parameters.
+
+```typescript
+export function traverseType(c : TreeCursor, s : string) : Type {
+  if(c.type.name === "MemberExpression") {
+    // parse generic type with type parameters
+    ...
+  } else {
+    ... 
+  }
+}
+```
+
+### Generic Class Definitions
+
+Parser should look for classes inheriting from the special Generic superclass and parse the type parameters.
+
+```typescript
+export function traverseClass(c : TreeCursor, s : string) : Class<null> {
+  const fields : Array<VarInit<null>> = [];
+  const methods : Array<FunDef<null>> = [];
+  const typeParams: Array<string> = [];
+  c.firstChild();
+  c.nextSibling(); // Focus on class name
+  const className = s.substring(c.from, c.to);
+  c.nextSibling(); // Focus on arglist/superclass
+  // Parse the superclass list and look for Generic
+  // and parse the type parameters if found
+  ...
+  return {
+    name: className,
+    fields,
+    methods,
+    typeParams
+  };
+}
+```
+
+### \_\_ZERO\_\_ Literal
+
+Parser should properly parse the new __ZERO__ value as a literal.
+
+```typescript
+export function traverseLiteral(c : TreeCursor, s : string) : Literal {
+  switch(c.type.name) {
+    case "VariableName":
+      // check if variable is called __ZERO__ and return literal 
+    default:
+      throw new Error("Not literal")
+  }
+}
+```
+
+---
+
+## Proposed changes to the Type-Checker
+
+### Tracking Type Variables and Class Type Parameters in Global Environment
+
+```typescript
+export type GlobalTypeEnv = {
+  globals: Map<string, Type>,
+  functions: Map<string, [Array<Type>, Type]>,
+  // classes has a third item in the tuple to track the list of the class' type parameters
+  classes: Map<string, [Map<string, Type>, Map<string, [Array<Type>, Type], Array<string>>]>
+  // list of defined type variables
+  typevars: Map<string, TypeVar>,
+}
+```
+
+### Type Hierarchy and Comparison
+
+The below functions need to be modified to take into account
+type parameters and generic types. With type parameter
+constraints, these could get pretty complex and we have not
+fleshed out the exact details completely at the moment.
```

Comment:
I would ignore constraints and do exactly equality on type parameter names. The inheritance group will already add nontrivial subtyping, let's not give ourselves too much to take on.